### ISSUE TITLE
Change in overlapping overload behavior from 1.1.393 to 1.1.394

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#10014]
**Describe the bug**
pyright 1.1.394 incorrectly reports overlapping overloads that 1.1.393 did not report

**Code or Screenshots**

```python
from typing import Generic, Literal, Sequence, TypeVar, overload

TTT = TypeVar("TTT", int, float)

class Foo(Generic[TTT]):
    @overload
    def __new__(
        cls, data: Sequence[int], *, dtype: Literal["int"] = ...
    ) -> Foo[int]: ...
    @overload
    def __new__(
        cls, data: Sequence[float], *, dtype: Literal["float"] = ...
    ) -> Foo[float]: ...

```

With 1.1.393, python 3.12, default pyright settings, no error is reported.
With 1.1.394 (and presumably 1.1.396 once I can install it on command line), I get
```text
newover1.pyi:7:9 - error: Overload 1 for "__new__" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)
1 error, 0 warnings, 0 informations
```

I don't see why these are overlapping overloads.  Note that the `dtype` argument is required to trigger this - if you remove it (and the asterisk), then both 1.1.393 and 1.1.394 report overlapping overloads.  But the `dtype` arguments are different, even though they are optional keyword arguments.

**VS Code extension or command-line**

pyright command line

**Comment by erictraut** [comment]
The new behavior is correct. The change in behavior is due to a fix that corrected [a previous bug](https://github.com/microsoft/pyright/issues/9840).

These overloads are overlapping because they both include a default argument for `dtype`. Pyright previously detected this case for regular parameters, but it was missing the same check for keyword-only arguments.

Consider the following example:

```python
def test(x: Sequence[float]):
    f = Foo(x)

    # The correct return type here is ambiguous due to an overlapping overload
    reveal_type(f) # Foo[float] or Foo[int]?

test([1, 2, 3])
```

**Comment by Dr-Irv** [comment]
> Consider the following example:
> ```python
> def test(x: Sequence[float]):
>     f = Foo(x)
> 
>     # The correct return type here is ambiguous due to an overlapping overload
>     reveal_type(f) # Foo[float] or Foo[int]?
> 
> test([1, 2, 3])
> ```

I would argue here that the `reveal_type(f)` statement should have revealed `Foo[float]` because the type of the argument to the `Foo` constructor is `Sequence[float]`.  

The goal of the above simple example is to differentiate between
```python
reveal_type(Foo([1, 2, 3]))
reveal_type(Foo([1.1, 2.2, 3.3]))
```
which `pyright` does see as being `Foo[int]` and `Foo[float]` respectively.

I'm fine to add the `reportOverlappingOverloads` ignore statement in `pandas-stubs` to get around this (like we do elsewhere for both `mypy` and `pyright`) but I don't see this as being overlapping.

**Comment by erictraut** [comment]
Yeah, the definition of "overlapping overload" isn't standardized. I recently proposed a definition for it for the typing spec. It turned out to be difficult to get consensus on that topic, so I [removed this section from the most recent draft chapter on overloads](https://github.com/python/typing/pull/1839/commits/e835221469f881e40d0948c14c684fa09417d8e0).

I don't have a strong opinion about how "overlapping overload" is defined as long as the definition is clear, principled, internally consistent, and consistent with type theory. Mypy tends to use a loose set of heuristics, which leads to difficult-to-explain inconsistencies. Pyright employs a clear, simple, and consistent set of rules. (Or rather, it tries to. The recent bug fix addressed a hole in its implementation.) But pyright's rules admittedly flag some overloads as overlapping in situations that users find annoying.

If you have some thoughts on a standardized definition for "overlapping overload" rules, you're welcome to propose that in typing forum. I'd be supportive of such an effort.

**Comment by Dr-Irv** [comment]
> If you have some thoughts on a standardized definition for "overlapping overload" rules, you're welcome to propose that in typing forum. I'd be supportive of such an effort.

I'm not the typing expert here, so not sure what to say.  What's interesting here is that while `pyright` reports the example below as having overlapping overloads, when we test the types, things are fine.

```python
from typing import Generic, Sequence, TypeVar, overload

TTT = TypeVar("TTT", int, float)

class Foo(Generic[TTT]):
    @overload
    def __new__(
        cls, data: Sequence[int]
    ) -> Foo[int]: ...
    @overload
    def __new__(
        cls, data: Sequence[float]
    ) -> Foo[float]: ...
```
`mypy` doesn't see this as overlapping, but `pyright` does.  If a user calls `Foo([1,2,3]`, the first overload is matched.  If a user calls `Foo([1.1., 1.2, 1.3])`, the second overload is matched.  So the behavior when you use the class works as intended, but `pyright` considers this to be overlapping, and I see that `Sequence[int]` is a subtype of `Sequence[float]` because `int` is a subtype of `float`.  I don't fully understand why `pyright` sees the above as being overlapping.

**Comment by erictraut** [comment]
Yes, when a user calls the `Foo` constructor directly with an argument of `[1, 2, 3]`, the first overload is matched. But I've provided an example above where the constructor is called indirectly with the same argument value of `[1, 2, 3]`, and the second overload is matched. That inconsistency is caused by an overlapping overload. Maybe you don't care about this particular case, but there are other cases where this inconsistency can cause problems. I think it's arguably a good idea for a type checker to report the potential problem. Mypy attempts to distinguish between different flavors of overlapping overload and report some but not report others. It's not clear to me what algorithm or heuristics they use. It's certainly not well documented, and its behavior strikes me as inconsistent.

**Comment by Dr-Irv** [comment]
> But I've provided an example above where the constructor is called indirectly with the same argument value of `[1, 2, 3]`, and the second overload is matched.

And I think that is the correct behavior.  The method `test(x: Sequence[float])` says it is expecting a `Sequence[float]`, so `Foo(x)` should be `Foo[float]`.  

>  I think it's arguably a good idea for a type checker to report the potential problem.

Sure, but IMHO, this isn't a potential problem.  If you declare 2 overloads and the first overload has an argument that is a subtype of the argument of the second overload, that's not overlapping.  

On the other hand, if you had something like
```python
@overload 
def bad_idea(x: list | tuple) -> str: ...
@overload
def bad_idea(x: list | dict) -> int: ...
```
that's definitely an overlapping overload because `list | tuple` is not a subtype of `list | dict`

Since we've established that the overloads get matched in sequential order, to me an overlapping overload is when this rule is true:
- Let $A$ be the set of types of arguments of the first overload
- Let $B$ be the set of types of arguments of the later overload
- Then the two overloads do not overlap if any of the following is true:
  - $A\cap B=\emptyset$, e.g., there is no intersection
  - $A\subset B$ and $B\setminus A \neq\emptyset$, e.g., $A$ is a subset of $B$ but not the same set.

Now, TBH, I am nowhere near as versed in type theory as you are, so maybe my rules don't make sense, but at least for me, they are conceptually easy to understand.

**Comment by jorenham** [comment]
I'm a big fan of set theory; so it's always good to see some in the wild :)

Couple of technical nitpicks:

- The strict subset relation $A \subset B$ already implies $B \setminus A \neq \emptyset$ (as opposed to $A \subseteq B$).
- Your definition does not cover the case where $A \not\subset B$ yet $A \cap B \neq \emptyset$, i.e. partial overlap on both sides, which applies to your `bad_idea(x)` example.
- This notation fits can (also) be considered as set-theoretical, so there's no need for type-theory to make sense of it.

**Comment by Dr-Irv** [comment]
> Couple of technical nitpicks:
> 
> * The strict subset relation $A\subset B$ already implies $B\setminus A\neq \emptyset$ (as opposed to $A\subseteq B$).

Yes, I knew that, but wasn't sure if every reader would know it.

> * Your definition does not cover the case where $A\not\subset B$ yet $A\cap  B\neq\emptyset$,, i.e. partial overlap on both sides, which applies to your `bad_idea(x)` example.

Well, it does because I am just listing the conditions where overloads would not overlap.  So the condition above is outside those conditions, thus they would overlap

**Comment by jorenham** [comment]
> Yes, I knew that, but wasn't sure if every reader would know it.

👍🏻 

> Well, it does because I am just listing the conditions where overloads would not overlap. So the condition above is outside those conditions, thus they would overlap

It's just this valid situation currently isn't covered:

```py
from typing import overload

@overload
def bad_idea(x: str | None) -> str | None: ...
@overload
def bad_idea(x: bytes | None) -> object: ...
```

So in case you were going for a *complete* description (in the scope of fully static types, i.e. without `Any` stuff), then you'd also have to consider situations like these. 

But don't take my comments too seriously; I'm just nitpicking on some undotted "i"'s here, realizing that this is unrelated to the (imo fair) point you're trying to make. I did to mean to start flailing a red herring around.

**Comment by Dr-Irv** [comment]
> It's just this valid situation currently isn't covered:
> 
> from typing import overload
> 
> @overload
> def bad_idea(x: str | None) -> str | None: ...
> @overload
> def bad_idea(x: bytes | None) -> object: ...

I think that is a case where you do have overlapping overloads.  `bad_idea(None)` is ambiguous, and my rules catch that as overlapping.

**Comment by jorenham** [comment]
> > It's just this valid situation currently isn't covered:
> > from typing import overload
> > 
> > @overload
> > def bad_idea(x: str | None) -> str | None: ...
> > @overload
> > def bad_idea(x: bytes | None) -> object: ...
> 
> I think that is a case where you do have overlapping overloads. `bad_idea(None)` is ambiguous, and my rules catch that as overlapping.

Ah I see. But does that really matter?
If the input can be assigned to both parameters, then that means that it must also be assignable to their intersection, `(str | None) & (bytes | None)`, which reduces to `None`. Since intersection `&` is distributive w.r.t. implication (`->`), we should therefore also take the intersection on the RHS, giving us `(str | None) & object`, which reduces to `str | None`.

In this case, this result is precisely equivalent to the first result type. If this wouldn't have been the case, then the overlapping overloads are incompatible.

If we apply this to your `-> int` and `-> str` example when we pass it a `list`, then we'd end up with `int & str`, so `Never`. Clearly `Never` is incompatible with `str`, so the overlapping overloads are indeed incompatible.

**Comment by Dr-Irv** [comment]
> If the input can be assigned to both parameters, then that means that it must also be assignable to their intersection, `(str | None) & (bytes | None)`, which reduces to `None`. Since intersection `&` is distributive w.r.t. implication (`->`), we should therefore also take the intersection on the RHS, giving us `(str | None) & object`, which reduces to `str | None`.
> 
> In this case, this result is precisely equivalent to the first result type. If this wouldn't have been the case, then the overlapping overloads are incompatible.

I'm not sure about that.

When I look at those overloads, the parameter `None` could return `str | None` if it matches the first overload, but it could return any object (e.g., `int`) if it matches the second overload.   So that's why they overlap.

**Comment by schl3ck** [comment]
If this is not the right place for this question then please point me to the right place.

**Edit:** I've found a discussion with the exact same problem: #10914. Therefore you can ignore this

I have a similar situation and I don't know how I should specify the overloads to avoid the error:

```python

@overload
def some_func(repeat: int = ..., accepted: None = ...) -> int: ...
@overload
def some_func(repeat: int = ..., accepted: list[int] = ...) -> list[int]: ...

def some_func(repeat: int = 2, accepted: list[int] | None = None) -> int | list[int]:
  # some implementation
```

Every parameter has a default value and the function is used in multiple places with and without keyword arguments so I can't change the whole parameter definition. I've added the parameter `accepted` that changes the return type. I don't want to make it a keyword-only argument.

Specified like above it works as expected (even with the error) so that when `accepted` is not specified I get the first overload and when it is specified (and not `None`) I get the second overload.

I cannot remove the default value from the function so how can I specify the overloads without error?
